nameserver/nameserver.go

```diff
@@ -204,10 +204,30 @@ func (n *Nameserver) receiveGossip(msg []byte) (mesh.GossipData, mesh.GossipData
 	}
 
 	n.Lock()
+	isKnownPeer := n.isKnownPeer
+	n.Unlock() // Don't hold the lock over call out to code elsewhere
+	gossip.Entries.filter(func(e *Entry) bool {
+		return isKnownPeer(e.Origin)
+	})
+
+	n.Lock()
 	defer n.Unlock()
 
+	// Check entries claiming to originate from us against our current data
 	gossip.Entries.filter(func(e *Entry) bool {
-		return n.isKnownPeer(e.Origin)
+		if e.Origin == n.ourName {
+			if ourEntry, ok := n.entries.findEqual(e); ok {
+				if ourEntry.Version < e.Version ||
+					(ourEntry.Version == e.Version && ourEntry.Tombstone != e.Tombstone) {
+					// Take our version of the data, but make the version higher than the incoming
+					ourEntry.Version = e.Version + 1
```

Comment:
modifying ourEntry here means the subsequent merge won't treat the entry as changed, which delays convergence. It's also somewhat dirty.

I suggest copying the entry as is and only increasing the version on the copy.